covid_graphs/covid_graphs/bayesian.py

```diff
@@ -0,0 +1,88 @@
+# Code inspired by Vladimír Boža
+
+import random
+from pathlib import Path
+
+import click
+import click_pathlib
+import matplotlib.pyplot as plt
+import numpy as np
+import pymc3 as pm
+
+from . import country_report
+from .country_graph import CountryGraph, GraphType
+from .fit_atg_model import AtgModelFit
+from .formula import FittedFormula
+from .predictions import CountryPrediction, PredictionEvent
+
+
+def fit_bayesian_model(cases: np.ndarray):
+    base_length = 800
+    with pm.Model() as model:  # noqa: F841
+        days = np.arange(len(cases))
+
+        alpha = pm.Uniform("alpha", 1, 20)
+        shift = pm.Uniform("shift", 0, 40)
+        peak = pm.Uniform("peak", 30, 80)
+        mult = pm.Uniform("mult", -30, 20)
+
+        tg = pm.Deterministic("tg", peak / alpha)
+        x = pm.math.maximum(0, (days - shift) / tg)
+        x_prev = pm.math.maximum(0, (days - 1 - shift) / tg)
+        exp_cases = pm.Deterministic(
+            "exp_cases",
+            pm.math.exp(mult - x) * (x ** alpha) - pm.math.exp(mult - x_prev) * (x_prev ** alpha),
+        )
+
+        sigma = pm.HalfCauchy("sigma", beta=500)
+        # likelihood = pm.Normal("y", mu=exp_cases, sigma=sigma, observed=cases)
+        # likelihood = pm.Cauchy("y", alpha=exp_cases, beta=sigma, observed=cases)
+        likelihood = pm.Laplace("obs", mu=exp_cases, b=sigma, observed=cases)  # noqa: F841
+
+        step = pm.NUTS(target_accept=0.9)
+        start = pm.find_MAP()
+        trace = pm.sample(
+            2 * base_length, chains=4, cores=4, tune=base_length, start=start, step=step
+        )
+        trace = trace[base_length:]
+
+    pm.traceplot(trace)
+    plt.show()
+
+    alpha_t = trace["alpha"][-base_length:]
+    shift_t = trace["shift"][-base_length:]
+    tg_t = trace["tg"][-base_length:]
+    a_t = np.exp(trace["mult"][-base_length:]) * tg_t
+
+    return [
+        AtgModelFit(exp=alpha, tg=tg, a=a, t0=shift + 1)
+        for alpha, tg, a, shift in zip(alpha_t, tg_t, a_t, shift_t)
+    ]
+
+
+@click.command(help="COVID-19 country growth visualization")
+@click.argument("filename", required=True, type=click_pathlib.Path(exists=True))
+def calculate_posterior(filename: Path):
+    report = country_report.create_report(filename)
+
+    cutoff = 1
+    cases = np.diff(report.cumulative_active[:-cutoff])
```

Comment:
Cutoff -1 is a strange value, you use all the datapoints except the last one.

Reply:
It should be an optional command line parameter. I've changed the default value to 0.

Reply:
It should be `len(report.cumulative_active)` instead of `0` in my opinion.

Reply:
This is how it's used, so value of 0 is fine. But perhaps the name cutoff is not the best, since you misunderstood.
```
    used_length = len(report.dates) - cutoff
    cases = np.diff(report.cumulative_active[:used_length])
```

Reply:
Oh, you updated the code... No it is completely fine, just I though it stayed `cases = np.diff(report.cumulative_active[:-cutoff])`.